[Concurrency] Ban passing actor state via inout

etcwilde opened the pull request:
Passing actor state to async functions via inout parameters violates
automicity. This patch bans passing actor state via an inout parameter
to an async function.

New test case verifies that we ban it when calling normal async
functions, functions that are called as a value decl instead of a
function decl, an async method on self, an async method on another
actor, something that has a curried function, and however else you want
to pass it.

Resolves: rdar://70635479

etcwilde commented:
@swift-ci please test

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -631,6 +631,42 @@ void swift::checkActorIsolation(const Expr *expr, const DeclContext *dc) {
+        // an actor.
+        FunctionType *fnType = apply->getFn()->getType()->getAs<FunctionType>();
+        if (fnType && fnType->isAsync()) {
+          // Since we get different types for the arg list depending on
+          // whether the function is called with one parameter or multiple
+          // parameters, just collect the inout expressions so that we don't
+          // have extra special-casing
+          llvm::SmallVector<Expr *, 3> inoutArgs;
+          if (ParenExpr *arg = dyn_cast<ParenExpr>(apply->getArg())) {
+            if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg->getSubExpr()))
+              inoutArgs.push_back(inoutArg->getSubExpr());
+          } else if (TupleExpr *argList =
+                         dyn_cast<TupleExpr>(apply->getArg())) {
+            inoutArgs.reserve(argList->getNumElements());
+            for (Expr *arg : argList->getElements()) {
+              if (InOutExpr *inoutArg = dyn_cast<InOutExpr>(arg))
+                inoutArgs.push_back(inoutArg->getSubExpr());
+            }
```
We might have to look through `ParenExpr` within each argument (e.g., for `f((&self.state))`). I recommend moving the collection of `inout` parameters out into a helper function.

etcwilde replied:
`f((&self.state))` already emits an error, though it's about an extraneous `&`, which isn't exactly clear. I could just strip the parens off too, but if it's invalid does it matter?

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -631,6 +631,42 @@ void swift::checkActorIsolation(const Expr *expr, const DeclContext *dc) {
+          // We have a list of arguments that were passed inout to this
+          // function, check that they don't come from an actor class
+          for (Expr *arg : inoutArgs) {
+            if (LookupExpr *lookupArg = dyn_cast<LookupExpr>(arg)) {
```
This might miss the case where the inout expression is digging into a member of an actor-isolated member, e.g.,

```swift
struct Point { var x: Int; var y: Int }
actor class A {
  var point: Point
  func f(_: inout Int) async { }
  func g() async { f(&point.x) } // should diagnose this
}
```

I wonder if instead of calling checkMemberReference directly here, we could somehow record which expressions have inout expression parents, and have checkMemberReference "look up" to find those expressions.

etcwilde replied:
Alright, when I strip off the layered `LookupExpr`s, we correctly emit an error message. Do we want the diagnostic to point at the `x` that is actually being passed, or do we want to point at the `point` since that's what belongs to the actor? Right now it looks like this, which is not wrong, but maybe confusing since it's actually passing the `x` property, not `position`:

```
/tmp/test2.swift:10:22: error: actor-isolated property 'position' cannot be passed 'inout' to asynchronous function                                                                                            
    await modifyInt(&position.x)
                     ^
```
I could probably do a note pointing at `position`, while saying that `x` is the thing that can't be passed in.

I'm doing the `LookupExpr` stripping directly in the the look over the `inoutArgs`. I don't think there would be an issue stripping off the layers of `LookupExpr`s in `checkMemberReference` since all of them do something to the effect of `lookupExpr->getBase()`. The only one that looks different is in the `SelfApplyExpr` clause where it just calls `getArg()`. I'm not sure if that's what you mean't with "having `checkMemberReference` "look up" to find the expressions"?

etcwilde commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -631,6 +646,47 @@ void swift::checkActorIsolation(const Expr *expr, const DeclContext *dc) {
+            // Strip off all but the last LookupExpr before the DeclRefExpr.
+            // If the Decl from that LookupExpr comes from the actor, it's a
+            // problem.
+            while (LookupExpr *lookupArg = dyn_cast<LookupExpr>(arg)) {
```
We could try to move this lookup into `checkMemberReference`, or maybe update `findMemberReference` to use it. I don't see a huge benefit to adding the loop to `checkMemberReference` since we'd need to run this loop in the caller as well to get the right `ConcreteDeclRef` to the member to pass into `checkMemberReference.

etcwilde commented:
@swift-ci please test

etcwilde committed:
[Concurrency] Ban passing actor state via inout

Passing actor statte to async functions via inout parameters violates
automicity. This patch bans passing actor state via inout parameter to
async functions.

This removes the warning when passing a sub-property of a property of an
actor class and replaces it with an error message since passing actor
state inout is not allowed.

etcwilde commented:
@swift-ci please smoke-test

kavon commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -578,6 +587,61 @@ findMemberReference(Expr *expr) {
+  // application.
+  auto extractInoutArgs = [&outputs](const ApplyExpr *call,
+                                     bool isPartialApply=false) {
+    if (ParenExpr *arg = dyn_cast<ParenExpr>(call->getArg())) {
```
I worry about code like `callee(((arg)))` here. You might want to skip this ParenExpr dyn_cast and use `callee->getArg()->getSemanticsProvidingExpr()` to look through arbitrary paren nesting. That should then allow you to simplify this if-else block a bit.

kavon commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -578,6 +587,61 @@ findMemberReference(Expr *expr) {
+  // isPartialApply should be true if the parameter is part of the partial
+  // application.
+  auto extractInoutArgs = [&outputs](const ApplyExpr *call,
+                                     bool isPartialApply=false) {
```
Nitpick: I think the default argument here isn't needed since the scope is so limited and both call-sites pass the argument explicitly. I also would suggest the formatting convention I've seen in LLVM where true/false literals passed in a call are annotated with the parameter's label, e.g., `extractInoutArgs(call, /*isPartialApply=*/false);` so that it's clear at the call-site what false means.

etcwilde replied:
- [ ] Remove default argument

kavon commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -578,6 +587,61 @@ findMemberReference(Expr *expr) {
+  // If part of the partial application takes something inout, then we should
+  // track it. A struct-type property of an actor that calls a mutating func is
+  // implicitly passed inout to the mutating function.
+  if (ApplyExpr * subCall = dyn_cast<ApplyExpr>(call->getFn()))
```
Nitpick: I would suggest being consistent with the `*` and either put it next to the type name, or the variable identifier, but not between the two like this. You seem to like the latter so I'd do that.

kavon commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -796,6 +864,54 @@ namespace {
+
+        auto memberDecl = baseArg->getMember().getDecl();
+        auto isolation = ActorIsolationRestriction::forDeclaration(memberDecl);
+        if (isolation == ActorIsolationRestriction::ActorSelf) {
```
I would suggest using an exhaustive switch (no `default` case) instead of this `if` test for actor isolation kind. The benefit of a switch is that if we end up adding more cases in the future, the compiler will point-out all of the places that need updating.

kavon commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -796,6 +864,54 @@ namespace {
+            // really only happen when the property is a struct with a mutating
+            // async method.
+            if (auto partialApply = dyn_cast<ApplyExpr>(call->getFn())) {
+              ValueDecl *fnDecl = cast<DeclRefExpr>(partialApply->getFn())->getDecl();
```
Nitpick: I think this line is over 80 col. If it is, add some new-lines

kavon commented on test/Concurrency/actor_inout_isolation.swift:
```diff
@@ -0,0 +1,105 @@
+  // Can't pass actor-isolated primitive into first-class function value
+  func inoutAsyncValueCall() async {
+    // expected-error@+1{{actor-isolated property 'value1' cannot be passed 'inout' to asynchronous function}}
+    await modifyAsyncValue(&value1)
```
Might be worth adding another call-site into this method that tests a closure application, e.g., `{ (_ foo: inout Int) async in foo += 1 }(&value1)` to cover that part of the AST.

etcwilde replied:
added in the `inoutAsyncClosureCall` test case.

kavon reviewed (commented):
I think it looks good. I added some style nitpicks and suggestions.

Before approving this, I wonder what happens currently for "implicitly async" call-site, such as:

```swift
extension TestActor {
  func syncModifyByValue(_ other: inout Int) { other += value1 }

  func modifyOther(_ ta : TestActor) async {
    await ta.syncModifyByValue(&value2)  // <- raise error for passing value2 via inout
  }
}
```

During TypeCheckConcurrency, the implicitly-async call-sites are marked (on the ApplyExpr), so there could be an awkward ordering problem for these `inout` checks.

etcwilde commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -537,6 +537,15 @@ namespace {
+  struct InOutArgInfo {
+    InOutExpr* arg;
+    bool isPartialApply; // base.foo() where foo is a mutating func on base struct
+  };
```
I'd agree, though maybe we could come up with a more descriptive comment for a Doxygen description?

etcwilde commented:
I can get rid of a lot of the AST decomposition now that the apply stack exists. When we hit an inout, I can just look at the top call in the apply stack and see if that is async. If it is, we can emit an error. This results in a smaller change that makes a bit more sense. Update coming soon.

etcwilde commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -578,6 +587,61 @@ findMemberReference(Expr *expr) {
+static bool isAsyncCall(const ApplyExpr *call) {
+  // Effectively the same as doing a
+  // `dyn_cast_or_null<FunctionType>(call->getFn()->getType())`, check the
+  // result of that and then checking `isAsync` if it's defined.
```
I put a more descriptive comment on the function itself. I could remove this one, but coming from LLVM, I kept looking for something like dyn_cast_or_null, so this was me reasoning about what was going on here (and thus was helpful and maybe could be helpful for someone else transitioning).

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci Please smoke test OS X platform

kavon commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -853,6 +886,58 @@ namespace {
+        case ActorIsolationRestriction::Unrestricted:
+        case ActorIsolationRestriction::LocalCapture:
+        case ActorIsolationRestriction::Unsafe:
+        case ActorIsolationRestriction::GlobalActor: // TODO: handle global
```
Agreed. I think we should make sure global actors are covered too.

kavon reviewed (commented):
Unless if I missed it, I think we may also want to have coverage for trying to pass state in an implicitly-async call. It's not currently well-described in the proposals, but "implicitly-async" means that we're allowing calls to actor-isolated sync functions from outside of that actor's context, but those call-sites are treated as async (and thus need an 'await'). I was thinking of something like this to cover it for passing to another actor-instance:

```swift
class TestActor {
  var state : Int
  // ...
  func passIt(_ other : DifferentActor) async {
    await other.modify(&state)  // <- implicitly-async call
  }
}

actor class DifferentActor {
  func modify(_ state : inout Int) { /**/ } // <- note this is sync
}
```

etcwilde commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -853,6 +886,58 @@ namespace {
+        case ActorIsolationRestriction::Unrestricted:
+        case ActorIsolationRestriction::LocalCapture:
+        case ActorIsolationRestriction::Unsafe:
+        case ActorIsolationRestriction::GlobalActor: // TODO: handle global
```
I have the patch for this, but I'd like to get this PR in before it grows too much bigger.

kavon replied:
Ah okay. That's fine with me

etcwilde commented:
> Unless if I missed it, I think we may also want to have coverage for trying to pass state in an implicitly-async call. It's not currently well-described in the proposals, but "implicitly-async" means that we're allowing calls to actor-isolated sync functions from outside of that actor's context, but those call-sites are treated as async (and thus need an 'await'). I was thinking of something like this to cover it for passing to another actor-instance:
> 
> ```swift
> class TestActor {
>   var state : Int
>   // ...
>   func passIt(_ other : DifferentActor) async {
>     await other.modify(&state)  // <- implicitly-async call
>   }
> }
> 
> actor class DifferentActor {
>   func modify(_ state : inout Int) { /**/ } // <- note this is sync
> }
> ```

Hmm, looks like I forgot to include this in the in-tree test case in the patch. Will fix!

etcwilde committed:
Adding test case for implicit async funcs

This patch adds a test to verify that functions that are implicitly
async also are reported as errors when passing data inout.

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -604,6 +604,23 @@ findMemberReference(Expr *expr) {
+  Type funcTypeType = call->getFn()->getType();
+  if (!funcTypeType)
+    return false;
+  FunctionType *funcType = funcTypeType->getAs<FunctionType>();
```
I think this could be converted into a `castTo<FunctionType>()` because otherwise it would never be a correct argument application (I'm assuming that this is checked after expression has been successfully type-checked).

etcwilde replied:
Changed

etcwilde committed:
Improve implicit async error message

At the declaration, the implicitly async functions appear to be
synchronous, so it isn't clear why the error message is being emitted.
This patch updates the error message to indicate that the function is
implicitly asynchronous.

etcwilde commented:
@swift-ci please smoke test this

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci Please smoke test Windows platform

kavon commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -4206,6 +4206,13 @@ ERROR(actor_isolated_self_independent_context,none,
+ERROR(actor_isolated_inout_state,none,
+      "actor-isolated %0 %1 cannot be passed 'inout' to"
+      "%select{| implicitly}2 asynchronous function",
+      (DescriptiveDeclKind, DeclName, bool))
```
I'd word this as `cannot be passed 'inout' %select{to asynchronous function|in an implicitly-asynchronous call}`, since it's only viewed as async from outside of actor contexts, and is otherwise synchronous inside of the actor. 

If it's possible, you may also want to emit the explainer Note diagnostic called `actor_isolated_sync_func` on the callee function (if it has a decl), which helps explain why the function is "implicitly asynchronous".

kavon replied:
Actually, why not just change the wording to be just `cannot be passed 'inout' in an asynchronous call` to cover both cases. Other parts of the typechecker will raise an error if the programmer forgot to mark the call as `await`, making the additional note just extra noise.

etcwilde replied:
I agree that specifying `call` probably makes sense since that's what's async. 

I'm a firm believer that if an error occurs due to something that is happening implicitly (even with the await), it should be noted that it is something being done implicitly. This avoids the programmer going back to the functionDecl and saying "hey, wait, I didn't label this async". Something explaining why it is implicitly async would definitely be a good thing to add though.

kavon replied:
The main reason why I backtracked on emitting the note here in TypeCheckConcurrency is that later on in TypeCheckEffects, we will emit that same note if the programmer forgot to cover that call-site with an `await`, which is a direct indicator that the programmer doesn't know that it's an implicitly async call-site.

I'm fine with emitting that note here in TypeCheckConcurrency for this error too, so feel free to do so if you'd like to. :)

etcwilde replied:
Ah, I wasn't quite sure which part you were referring to as being extra noise. If other parts are emitting the explanation on why it's async, that's fine by me, I won't add it here. I'll just add the bit saying that it's the async call.

kavon reviewed (approved):
Ok it LGTM now, covering everything I can think of. Just added one additional suggestion about the message and possibly adding a note (if you're able to do so at the point of the diagnostic).

etcwilde commented:
@swift-ci please smoke test
